File: Data Structures & Algorithms/binary-tree-right-side-view/submission-4.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun rightSideView(root: TreeNode?): List<Int> {
        val result = mutableListOf<Int>()
        val stack = ArrayDeque<TreeNode>()

        root?.let { stack.add(it) }

        while (stack.size > 0) {
            val inner = mutableListOf<Int>()
            for (i in 0 until stack.size) {
                val current = stack.removeFirst()
                inner.add(current.`val`)
                current.left?.let { stack.add(it) }
                current.right?.let { stack.add(it) }
            }
            if (inner.size > 0) {
                result.add(inner.last())
            }
        }

        return result
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        current = self.findDepth(root.left) + self.findDepth(root.right)
        left = self.diameterOfBinaryTree(root.left)
        right = self.diameterOfBinaryTree(root.right)

        return max(current, left, right)

    def findDepth(self, root: Optional[TreeNode], depth = 0):
        if not root:
            return depth

        return max(self.findDepth(root.left, depth + 1), self.findDepth(root.right, depth + 1))